File: predicates.py
"""
predicates.py — Typed post-conditions for plan steps.

A predicate is a structured, deterministically-checkable assertion about
desktop / file / browser state that a plan step claims will hold once it
finishes.  Promoting `expected` from free text to a typed predicate lets
the controller verify a step's success without an LLM round-trip and
fire ``PREDICATE_NOT_MET`` failure classification automatically when it
doesn't hold.

Predicate kinds
---------------
  window_title_contains   — any visible window title contains a substring
  window_active_app       — currently focused window matches an app name
  file_exists             — a path resolves to an existing file
  file_absent             — a path does NOT resolve to a file
  file_contains           — a file's body contains a substring
  url_contains            — the browser's current URL contains a substring
  text_visible            — a visible label is on screen (a11y or OCR)
  process_running         — a process matching a name pattern is running
  shell_returns           — a probe shell command exits 0 with optional stdout match

The controller reads ``step.predicate`` (when present) after the step's
``STEP_DONE`` marker fires and calls ``check_predicate``.  A predicate
miss converts the verdict to ``BLOCKED`` so the standard replan / retry
machinery kicks in.

Predicates are stored as plain JSON-serialisable dicts so they survive
the Plan.to_dict() round-trip and replay.
"""

from __future__ import annotations

import asyncio
import json
import logging
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any

logger = logging.getLogger(__name__)


@dataclass
class PredicateResult:
    ok: bool
    kind: str
    detail: str = ""
    observed: Any = None


# ---------------------------------------------------------------------------
# Validation
# ---------------------------------------------------------------------------

_KNOWN_KINDS = frozenset({
    "window_title_contains",
    "window_active_app",
    "file_exists",
    "file_absent",
    "file_contains",
    "url_contains",
    "text_visible",
    "process_running",
    "shell_returns",
})


def normalize(predicate: Any) -> dict | None:
    """
    Coerce a free-form predicate spec into a canonical dict.  Returns
    None when the input is not a usable predicate (so callers can fall
    back to LLM-judged ``expected`` text).
    """
    if not isinstance(predicate, dict):
        return None
    kind = predicate.get("kind") or predicate.get("type")
    if kind not in _KNOWN_KINDS:
        return None
    out = dict(predicate)
    out["kind"] = kind
    out.pop("type", None)
    return out


def render(predicate: dict) -> str:
    """One-line human-readable description for prompts and traces."""
    kind = predicate.get("kind", "?")
    if kind == "window_title_contains":
        return f"a window with title containing {predicate.get('value')!r}"
    if kind == "window_active_app":
        return f"the focused window's app matches {predicate.get('value')!r}"
    if kind == "file_exists":
        return f"file exists: {predicate.get('path')!r}"
    if kind == "file_absent":
        return f"file does NOT exist: {predicate.get('path')!r}"
    if kind == "file_contains":
        return f"file {predicate.get('path')!r} contains {predicate.get('value')!r}"
    if kind == "url_contains":
        return f"browser URL contains {predicate.get('value')!r}"
    if kind == "text_visible":
        return f"text visible on screen: {predicate.get('value')!r}"
    if kind == "process_running":
        return f"process matching {predicate.get('value')!r} is running"
    if kind == "shell_returns":
        cmd = predicate.get("command", "")
        m = predicate.get("stdout_contains")
        return f"`{cmd[:60]}` exits 0" + (f" and stdout contains {m!r}" if m else "")
    return kind


# ---------------------------------------------------------------------------
# Checking
# ---------------------------------------------------------------------------

async def check_predicate(predicate: dict, registry) -> PredicateResult:
    """
    Evaluate ``predicate`` against the live system using ``registry`` for
    tool dispatch.  Never raises — failure modes return ``ok=False`` with
    an explanatory ``detail``.

    The registry is the standalone-agent ToolRegistry; this function calls
    desktop_list_windows / desktop_get_active_window / fs_read /
    desktop_find_text / browser_eval / shell_run as appropriate.
    """
    p = normalize(predicate)
    if p is None:
        return PredicateResult(False, "?", "predicate normalised to None")
    kind = p["kind"]

    try:
        if kind == "window_title_contains":
            return await _check_window_title(p, registry)
        if kind == "window_active_app":
            return await _check_active_app(p, registry)
        if kind == "file_exists":
            return _check_file_presence(p, must_exist=True)
        if kind == "file_absent":
            return _check_file_presence(p, must_exist=False)
        if kind == "file_contains":
            return _check_file_contains(p)
        if kind == "url_contains":
            return await _check_url(p, registry)
        if kind == "text_visible":
            return await _check_text_visible(p, registry)
        if kind == "process_running":
            return await _check_process(p, registry)
        if kind == "shell_returns":
            return await _check_shell(p, registry)
    except Exception as e:
        logger.warning("[predicates] check %s raised: %s", kind, e)
        return PredicateResult(False, kind, f"check raised: {e}")

    return PredicateResult(False, kind, "no checker for this kind")


# ---------------------------------------------------------------------------
# Per-kind checkers
# ---------------------------------------------------------------------------

async def _check_window_title(p: dict, registry) -> PredicateResult:
    needle = str(p.get("value") or "")
    if not needle:
        return PredicateResult(False, p["kind"], "empty value")
    if "desktop_list_windows" not in set(registry.list_tools()):
        return PredicateResult(False, p["kind"], "backend has no list_windows")
    raw = await registry.dispatch("desktop_list_windows", {})
    try:
        result = json.loads(raw)
    except (json.JSONDecodeError, TypeError):
        return PredicateResult(False, p["kind"], "window listing unparsable")
    # Surface dispatch-level errors directly so the model sees a clear
    # diagnostic instead of "no window contains …" when the real issue
    # is "tool returned an error".
    if isinstance(result, dict) and "error" in result:
        return PredicateResult(False, p["kind"],
                               f"list_windows error: {result['error']}",
                               result)
    wins = result.get("windows") if isinstance(result, dict) else []
    wins = wins or []
    for w in wins:
        title = str(w.get("title", "") or "")
        if needle.lower() in title.lower():
            return PredicateResult(True, p["kind"], f"matched {title!r}", w)
    return PredicateResult(False, p["kind"],
                           f"no window title contains {needle!r}",
                           {"windows": [w.get("title", "") for w in wins]})


async def _check_active_app(p: dict, registry) -> PredicateResult:
    target = str(p.get("value") or "").lower()
    if not target:
        return PredicateResult(False, p["kind"], "empty value")
    if "desktop_get_active_window" not in set(registry.list_tools()):
        return PredicateResult(False, p["kind"], "backend has no get_active_window")
    raw = await registry.dispatch("desktop_get_active_window", {})
    try:
        info = json.loads(raw)
    except json.JSONDecodeError:
        return PredicateResult(False, p["kind"], "active-window listing unparsable")
    if not info or not info.get("found"):
        return PredicateResult(False, p["kind"], "no active window detected")
    win = info.get("window") or info
    app = str(win.get("app", "") or "").lower()
    title = str(win.get("title", "") or "").lower()
    if target in app or target in title:
        return PredicateResult(True, p["kind"], f"active app/title matched {target!r}", win)
    return PredicateResult(False, p["kind"],
                           f"active app={app!r} title={title!r}", win)


def _check_file_presence(p: dict, *, must_exist: bool) -> PredicateResult:
    path = str(p.get("path") or "")
    if not path:
        return PredicateResult(False, p["kind"], "empty path")
    # ``file_exists`` / ``file_absent`` are *file* predicates — a directory
    # at the path should NOT satisfy file_exists, and a directory should
    # NOT make file_absent fail.  Use is_file so both cases ignore
    # directories the same way callers naturally read the kind name.
    expanded = Path(path).expanduser()
    is_file = expanded.is_file()
    if is_file == must_exist:
        return PredicateResult(True, p["kind"],
                               f"file_{'exists' if is_file else 'absent'} satisfied", path)
    # Distinguish "missing" from "is a directory" so the failure detail
    # tells the caller which of those is actually true; a generic
    # "unexpectedly absent" message hides the latter case.
    if must_exist:
        if expanded.is_dir():
            detail = f"path is a directory, not a file: {path}"
        else:
            detail = f"file not found: {path}"
    else:
        # must_exist=False but is_file=True
        detail = f"file unexpectedly present: {path}"
    return PredicateResult(False, p["kind"], detail)


def _check_file_contains(p: dict) -> PredicateResult:
    path = str(p.get("path") or "")
    needle = str(p.get("value") or "")
    if not path or not needle:
        return PredicateResult(False, p["kind"], "empty path or value")
    f = Path(path).expanduser()
    if not f.is_file():
        return PredicateResult(False, p["kind"], f"file missing: {path}")
    try:
        body = f.read_text(encoding="utf-8", errors="replace")
    except OSError as e:
        return PredicateResult(False, p["kind"], f"read failed: {e}")
    if needle in body:
        return PredicateResult(True, p["kind"], "substring present", path)
    # Try regex if the value looks like one.
    if needle.startswith("/") and needle.endswith("/") and len(needle) > 2:
        pat = needle[1:-1]
        try:
            if re.search(pat, body):
                return PredicateResult(True, p["kind"], "regex matched", path)
        except re.error:
            pass
    return PredicateResult(False, p["kind"],
                           f"substring {needle!r} not in file ({len(body)}B)")


async def _check_url(p: dict, registry) -> PredicateResult:
    needle = str(p.get("value") or "")
    if not needle:
        return PredicateResult(False, p["kind"], "empty value")
    if "browser_eval" not in set(registry.list_tools()):
        return PredicateResult(False, p["kind"],
                               "browser tools unavailable; cannot verify URL")
    raw = await registry.dispatch("browser_eval", {"expression": "window.location.href"})
    try:
        result = json.loads(raw)
    except json.JSONDecodeError:
        return PredicateResult(False, p["kind"], "browser_eval result unparsable")
    if "error" in result:
        return PredicateResult(False, p["kind"], f"browser_eval error: {result['error']}")
    href = str(result.get("value", "") or "")
    if needle.lower() in href.lower():
        return PredicateResult(True, p["kind"], f"URL matched {href}", href)
    return PredicateResult(False, p["kind"],
                           f"URL {href!r} does not contain {needle!r}", href)


async def _check_text_visible(p: dict, registry) -> PredicateResult:
    needle = str(p.get("value") or "")
    if not needle:
        return PredicateResult(False, p["kind"], "empty value")
    tools = set(registry.list_tools())
    if "desktop_find_text" in tools:
        raw = await registry.dispatch("desktop_find_text", {"text": needle})
        try:
            result = json.loads(raw)
        except json.JSONDecodeError:
            result = {}
        if result.get("found"):
            return PredicateResult(True, p["kind"], f"text {needle!r} visible",
                                   result.get("match"))
        return PredicateResult(False, p["kind"], f"text {needle!r} not on screen", result)
    return PredicateResult(False, p["kind"],
                           "no OCR backend available for text-visible check")


async def _check_process(p: dict, registry) -> PredicateResult:
    needle = str(p.get("value") or "")
    if not needle:
        return PredicateResult(False, p["kind"], "empty value")
    if "shell_run" not in set(registry.list_tools()):
        return PredicateResult(False, p["kind"], "shell unavailable; cannot list processes")
    import platform as _p
    cmd = (
        f'tasklist /FI "IMAGENAME eq {needle}*"' if _p.system() == "Windows"
        else f'pgrep -lf {needle!r}'
    )
    raw = await registry.dispatch("shell_run", {"command": cmd})
    try:
        result = json.loads(raw)
    except json.JSONDecodeError:
        return PredicateResult(False, p["kind"], "shell result unparsable")
    out = (result.get("stdout") or "")
    if needle.lower() in out.lower():
        return PredicateResult(True, p["kind"], "process found", out[:160])
    return PredicateResult(False, p["kind"],
                           f"no process matching {needle!r} found", out[:160])


async def _check_shell(p: dict, registry) -> PredicateResult:
    cmd = str(p.get("command") or "")
    expect = p.get("stdout_contains")
    if not cmd:
        return PredicateResult(False, p["kind"], "empty command")
    if "shell_run" not in set(registry.list_tools()):
        return PredicateResult(False, p["kind"], "shell unavailable")
    raw = await registry.dispatch("shell_run", {"command": cmd})
    try:
        result = json.loads(raw)
    except json.JSONDecodeError:
        return PredicateResult(False, p["kind"], "shell result unparsable")
    if result.get("exit_code") not in (0, None):
        return PredicateResult(False, p["kind"],
                               f"exit_code={result.get('exit_code')}",
                               result.get("stderr", "")[:160])
    if expect:
        out = result.get("stdout") or ""
        if str(expect) not in out:
            return PredicateResult(False, p["kind"],
                                   f"stdout missing {expect!r}", out[:160])
    return PredicateResult(True, p["kind"], "shell probe satisfied",
                           (result.get("stdout") or "")[:160])


# ---------------------------------------------------------------------------
# Helper for offline (unit-test) predicate evaluation.
# ---------------------------------------------------------------------------

def check_filesystem_predicate_sync(predicate: dict) -> PredicateResult:
    """Subset of check_predicate that doesn't need a registry — for tests."""
    p = normalize(predicate)
    if p is None:
        return PredicateResult(False, "?", "invalid predicate")
    if p["kind"] == "file_exists":
        return _check_file_presence(p, must_exist=True)
    if p["kind"] == "file_absent":
        return _check_file_presence(p, must_exist=False)
    if p["kind"] == "file_contains":
        return _check_file_contains(p)
    return PredicateResult(False, p["kind"], "needs registry; use check_predicate")


__all__ = ["PredicateResult", "normalize", "render", "check_predicate",
           "check_filesystem_predicate_sync"]

File: pi-extension/src/predicates.ts
/**
 * predicates.ts — Typed post-conditions for plan steps.
 *
 * Mirror of Python ``predicates.py``.  The pi extension uses these
 * primarily as a planning vocabulary the model writes into the typed
 * plan, so the standalone agent (mainline) can verify them
 * deterministically when the same plan replays through its
 * controller.  The extension also exposes a ``check_predicate`` tool
 * for in-loop verification when Pi's agent wants to verify a
 * post-condition explicitly.
 */

import { existsSync, readFileSync, statSync } from "node:fs";
import type { DesktopBackend } from "./types.js";

function isRegularFile(path: string): boolean {
  // Match Python predicates._check_file_presence: directories must NOT
  // satisfy file_exists and must NOT make file_absent fail.  Use stat
  // and gracefully degrade to "not a file" on any I/O error.
  try {
    return statSync(path).isFile();
  } catch {
    return false;
  }
}

function shellQuote(s: string): string {
  // POSIX shell single-quote escape: surround with single quotes and
  // replace any embedded ' with '\'' (close-quote, escaped-quote,
  // re-open-quote).  Mirrors Python's shlex.quote for non-empty input.
  return `'${s.replace(/'/g, "'\\''")}'`;
}

export const PREDICATE_KINDS = [
  "window_title_contains",
  "window_active_app",
  "file_exists",
  "file_absent",
  "file_contains",
  "url_contains",
  "text_visible",
  // Model-generated synonym — handled as text_visible.  Mirrors the
  // mainline alias in predicates.py.
  "text_in_window",
  "process_running",
  "shell_returns",
] as const;

export type PredicateKind = (typeof PREDICATE_KINDS)[number];

export interface Predicate {
  kind: PredicateKind;
  value?: string;
  path?: string;
  command?: string;
  stdout_contains?: string;
}

export interface PredicateResult {
  ok: boolean;
  kind: string;
  detail: string;
  observed?: unknown;
}

export function normalizePredicate(p: unknown): Predicate | undefined {
  if (!p || typeof p !== "object") return undefined;
  const obj = p as Record<string, unknown>;
  const kind = obj["kind"] ?? obj["type"];
  if (typeof kind !== "string" || !PREDICATE_KINDS.includes(kind as PredicateKind)) {
    return undefined;
  }
  const out: Predicate = { kind: kind as PredicateKind };
  for (const k of ["value", "path", "command", "stdout_contains"] as const) {
    const v = obj[k];
    if (typeof v === "string") (out as unknown as Record<string, unknown>)[k] = v;
  }
  return out;
}

export function renderPredicate(p: Predicate): string {
  switch (p.kind) {
    case "window_title_contains": return `a window with title containing ${JSON.stringify(p.value)}`;
    case "window_active_app":     return `the focused window's app matches ${JSON.stringify(p.value)}`;
    case "file_exists":           return `file exists: ${JSON.stringify(p.path)}`;
    case "file_absent":           return `file does NOT exist: ${JSON.stringify(p.path)}`;
    case "file_contains":         return `file ${JSON.stringify(p.path)} contains ${JSON.stringify(p.value)}`;
    case "url_contains":          return `browser URL contains ${JSON.stringify(p.value)}`;
    case "text_visible":
    case "text_in_window":        return `text visible on screen: ${JSON.stringify(p.value)}`;
    case "process_running":       return `process matching ${JSON.stringify(p.value)} running`;
    case "shell_returns": {
      const m = p.stdout_contains ? ` and stdout contains ${JSON.stringify(p.stdout_contains)}` : "";
      return `\`${(p.command ?? "").slice(0, 60)}\` exits 0${m}`;
    }
    default: return p.kind;
  }
}

/** Filesystem predicates that need no backend.  Useful in tests + offline checks. */
export function checkFilesystemPredicateSync(p: Predicate): PredicateResult {
  switch (p.kind) {
    case "file_exists": {
      const path = p.path ?? "";
      if (!path) return { ok: false, kind: p.kind, detail: "empty path" };
      if (isRegularFile(path)) {
        return { ok: true, kind: p.kind, detail: `path=${path}` };
      }
      const exists = existsSync(path);
      return {
        ok: false,
        kind: p.kind,
        detail: exists
          ? `path is a directory, not a file: ${path}`
          : `missing: ${path}`,
      };
    }
    case "file_absent": {
      const path = p.path ?? "";
      if (!path) return { ok: false, kind: p.kind, detail: "empty path" };
      // file_absent passes for both "no entry" and "directory at this path"
      // — only a regular file should count as "present".
      if (!isRegularFile(path)) {
        return { ok: true, kind: p.kind, detail: `path=${path}` };
      }
      return { ok: false, kind: p.kind, detail: `unexpectedly present: ${path}` };
    }
    case "file_contains": {
      const path = p.path ?? "";
      const needle = p.value ?? "";
      if (!path || !needle) return { ok: false, kind: p.kind, detail: "empty path or value" };
      try {
        const body = readFileSync(path, "utf8");
        return body.includes(needle)
          ? { ok: true, kind: p.kind, detail: "substring present" }
          : { ok: false, kind: p.kind, detail: `substring ${JSON.stringify(needle)} not in file` };
      } catch (e) {
        return { ok: false, kind: p.kind, detail: `read failed: ${(e as Error).message}` };
      }
    }
    default:
      return { ok: false, kind: p.kind, detail: "needs runtime; use checkPredicate" };
  }
}

/**
 * Live predicate check.  ``backend`` is required for desktop-related
 * checks; ``runShell`` is an optional shell-runner closure that the
 * caller (tools.ts) can supply when shell access is allowed.
 */
export async function checkPredicate(
  predicate: Predicate,
  backend: DesktopBackend | undefined,
  helpers: {
    runShell?: (cmd: string) => Promise<{ stdout: string; stderr: string; exitCode: number }>;
    browserEval?: (expr: string) => Promise<unknown>;
    findText?: (text: string) => Promise<{ found: boolean }>;
  },
): Promise<PredicateResult> {
  switch (predicate.kind) {
    case "file_exists":
    case "file_absent":
    case "file_contains":
      return checkFilesystemPredicateSync(predicate);
    case "window_title_contains": {
      if (!backend) return { ok: false, kind: predicate.kind, detail: "no backend" };
      const needle = (predicate.value ?? "").toLowerCase();
      if (!needle) return { ok: false, kind: predicate.kind, detail: "empty value" };
      try {
        const r = await backend.listWindows();
        const match = r.windows.find((w) => (w.title ?? "").toLowerCase().includes(needle));
        return match
          ? { ok: true, kind: predicate.kind, detail: `matched ${JSON.stringify(match.title)}`, observed: match }
          : { ok: false, kind: predicate.kind, detail: `no window contains ${JSON.stringify(predicate.value)}` };
      } catch (e) {
        return { ok: false, kind: predicate.kind, detail: `listWindows failed: ${(e as Error).message}` };
      }
    }
    case "window_active_app": {
      if (!backend) return { ok: false, kind: predicate.kind, detail: "no backend" };
      const target = (predicate.value ?? "").toLowerCase();
      if (!target) return { ok: false, kind: predicate.kind, detail: "empty value" };
      try {
        const info = await backend.activeWindow();
        if (!info.found || !info.window) return { ok: false, kind: predicate.kind, detail: "no active window" };
        const app = (info.window.app ?? "").toLowerCase();
        const title = (info.window.title ?? "").toLowerCase();
        return (app.includes(target) || title.includes(target))
          ? { ok: true, kind: predicate.kind, detail: `active app/title contains ${target}`, observed: info.window }
          : { ok: false, kind: predicate.kind, detail: `active app=${app} title=${title}`, observed: info.window };
      } catch (e) {
        return { ok: false, kind: predicate.kind, detail: `activeWindow failed: ${(e as Error).message}` };
      }
    }
    case "url_contains": {
      const needle = (predicate.value ?? "").toLowerCase();
      if (!needle) return { ok: false, kind: predicate.kind, detail: "empty value" };
      if (!helpers.browserEval) return { ok: false, kind: predicate.kind, detail: "browser unavailable" };
      try {
        const value = await helpers.browserEval("window.location.href");
        const href = String(value ?? "").toLowerCase();
        return href.includes(needle)
          ? { ok: true, kind: predicate.kind, detail: `URL ${href}`, observed: href }
          : { ok: false, kind: predicate.kind, detail: `URL ${href} does not contain ${needle}`, observed: href };
      } catch (e) {
        return { ok: false, kind: predicate.kind, detail: `browser_eval failed: ${(e as Error).message}` };
      }
    }
    case "text_visible":
    case "text_in_window": {
      const needle = predicate.value ?? "";
      if (!needle) return { ok: false, kind: predicate.kind, detail: "empty value" };
      if (!helpers.findText) return { ok: false, kind: predicate.kind, detail: "OCR unavailable" };
      try {
        const r = await helpers.findText(needle);
        return r.found
          ? { ok: true, kind: predicate.kind, detail: `text visible: ${needle}` }
          : { ok: false, kind: predicate.kind, detail: `text ${JSON.stringify(needle)} not on screen` };
      } catch (e) {
        return { ok: false, kind: predicate.kind, detail: `find_text failed: ${(e as Error).message}` };
      }
    }
    case "process_running": {
      if (!helpers.runShell) return { ok: false, kind: predicate.kind, detail: "shell unavailable" };
      const needle = predicate.value ?? "";
      if (!needle) return { ok: false, kind: predicate.kind, detail: "empty value" };
      // The needle reaches /bin/sh -c (or cmd /C) verbatim.  JSON.stringify
      // is a JS string literal, NOT shell-safe quoting — it leaves single
      // quotes / $() / backticks untouched and so allowed both broken
      // matching and command injection.  Single-quote escape on POSIX,
      // and on Windows reject obvious shell metacharacters since
      // tasklist's /FI takes a double-quoted string with no portable
      // escape for embedded double quotes.
      let cmd: string;
      if (process.platform === "win32") {
        if (/["&|<>%^]/.test(needle)) {
          return {
            ok: false, kind: predicate.kind,
            detail: `refusing to query process: needle ${JSON.stringify(needle)} contains shell metacharacters`,
          };
        }
        cmd = `tasklist /FI "IMAGENAME eq ${needle}*"`;
      } else {
        // `--` defangs a needle that starts with `-` so pgrep treats it
        // as a pattern instead of an unknown option.  shellQuote still
        // handles the value's metacharacters.  Mirrors the Python
        // predicates._check_process fix.
        cmd = `pgrep -lf -- ${shellQuote(needle)}`;
      }
      try {
        const r = await helpers.runShell(cmd);
        const found = r.stdout.toLowerCase().includes(needle.toLowerCase());
        return found
          ? { ok: true, kind: predicate.kind, detail: "process found" }
          : { ok: false, kind: predicate.kind, detail: `no process matching ${needle} found` };
      } catch (e) {
        return { ok: false, kind: predicate.kind, detail: `shell failed: ${(e as Error).message}` };
      }
    }
    case "shell_returns": {
      if (!helpers.runShell) return { ok: false, kind: predicate.kind, detail: "shell unavailable" };
      const cmd = predicate.command ?? "";
      if (!cmd) return { ok: false, kind: predicate.kind, detail: "empty command" };
      try {
        const r = await helpers.runShell(cmd);
        if (r.exitCode !== 0) return { ok: false, kind: predicate.kind, detail: `exit ${r.exitCode}` };
        if (predicate.stdout_contains && !r.stdout.includes(predicate.stdout_contains)) {
          return { ok: false, kind: predicate.kind, detail: `stdout missing ${predicate.stdout_contains}` };
        }
        return { ok: true, kind: predicate.kind, detail: "shell probe satisfied" };
      } catch (e) {
        return { ok: false, kind: predicate.kind, detail: `shell failed: ${(e as Error).message}` };
      }
    }
  }
}
